### DISCUSSION TITLE
V5 beta breaking change? createMachine<ContextType,EventType, StateType> complains about numbers of generic arguments!

### DISCUSSION THREAD

**Discussion by Simonl9l** [statelyai/xstate#4276] [General]
Is there the need for additional "breaking change" documentation for v4?

With v4 one could provide typescript types for the Context, Events and States. However I now get a Typescript error, that the declaration now requires 10 not 3 generic parameters?

I initially followed this [youtube](https://www.youtube.com/watch?v=7tT45eRmyJE&t=87s) vue3 tutorial -  where they declared these type outside.

I was looking to see if I can switch this to use the `types:` declaration per [this](https://stately.ai/docs/xstate-v5/machines#specifying-types), but it does _not_ seem to have an element for `states"`?

Also in the tutorial, one had the option to define a different context type for each state and leverage typescript intersection types.

```ts
type MyContext = {
  myContext: string | undefined;
  myOtherContext: int | undefined;
}

type OtherStateContext = MyContext & {
  myContext: string | undefined;
  myOtherContext: int;
}

type InitialState = { value:'Initial'; context: MyContext }
type OtherState = { value:'Other'; context: OtherStateContext }

type MyStates =
 | InitialState
 | OtherState;

type MyEvents =
  { type: "start" } |
  { type: "stop" } |
  { type: "etc };

export const myMachine = createMachine<MyContext, MyEvents, MyStates>({
  id: "my_machine"
...});
```

how is all this achievable in V5

As an asside I pinged the `xstate-pinia` folks to see if they were planning to support V5....  https://github.com/wobsoriano/pinia-xstate/issues/3#issuecomment-1717140926

**Comment by davidkpiano** [comment]
You should not have to provide _any_ generic types.

This should look like:

```ts
export const myMachine = createMachine({
  types: {} as {
    context: MyContext;
    events: MyEvents;
    // state types not supported right now
  },
  // ...
});
```

**Comment by Simonl9l** [comment]
@davidkpiano thanks (on multiple fronts - xstate-pinia! too) 

Ok I figured as much...(video was prob not the nest starting point even for v4 - it was a vue3 example as they seem fairly thin on the ground)!

How far out aprox. is `states:` support?

Thanks again!

**Reply by davidkpiano** [reply]
States support will likely be either Typegen unless we can figure out a way to make it work with normal TypeScript. @Andarist has been looking into this.

**Reply by Andarist** [reply]
There are multiple tiers of the potential `states` support. So far we bump into TS limitations when it comes to even providing the most basic one 😅 So I'd say that "typestates" support is pretty far away right now. We are constantly evaluating our options here and I'm trying the weirdest things to get something working but it just proves to be quite challenging.
